src/wayland/shell/xdg/mod.rs

```diff
@@ -108,36 +113,270 @@ mod xdg_handlers;
 // compatibility handlers for the zxdg_shell_v6 protocol, its earlier version
 mod zxdgv6_handlers;
 
-/// Metadata associated with the `xdg_surface` role
+macro_rules! xdg_role {
+    ($(#[$role_attr:meta])* $role:ident,
+     $configure:ty,
+     $(#[$attr:meta])* $element:ident {$($(#[$field_attr:meta])* $vis:vis$field:ident:$type:ty),*},
+     $role_ack_configure:expr) => {
+        $(#[$role_attr])*
+        pub type $role = Option<$element>;
+
+        $(#[$attr])*
+        pub struct $element {
+                /// Defines the current geometry of the surface if any.
+                pub geometry: Option<Rectangle>,
+
+                /// Holds the double-buffered geometry that may be specified
+                /// by xdg_surface.set_geometry. This should be moved to the
+                /// current configuration on a commit of the underlying
+                /// wl_surface.
+                pub next_geometry: Option<Rectangle>,
+
+                /// Defines if the surface has received at least one
+                /// xdg_surface.ack_configure from the client
+                pub configured: bool,
+
+                /// The serial of the last acked configure
+                pub configure_serial: Option<Serial>,
+
+                /// Holds the next serial that will be used for
+                /// xdg_surface.configure
+                pub configure_next_serial: Serial,
```

Comment:
This should just be using the global serial counter, we don't need to have independent serials for each surface, quite the contrary.

Reply:
I have removed this field and moved most of the configure logic to the `ToplevelSurface` and `PopupSurface`. This has the nice side effect of reducing code duplication between `xdg_wm_base` and `zgdxv6`.